stock_demand_estimate/wizards/stock_demand_estimate_wizard.py

```diff
@@ -0,0 +1,202 @@
+# Copyright 2016 Eficent Business and IT Consulting Services S.L.
+#   (http://www.eficent.com)
+# Copyright 2016 Aleph Objects, Inc. (https://www.alephobjects.com/)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+
+from odoo import api, fields, models, _
+import odoo.addons.decimal_precision as dp
+from odoo.exceptions import UserError, ValidationError
+
+
+class StockDemandEstimateSheet(models.TransientModel):
+    _name = 'stock.demand.estimate.sheet'
+    _description = 'Stock Demand Estimate Sheet'
+
+    def _default_estimate_ids(self):
+        date_start = self.env.context.get('default_date_start', False)
+        date_end = self.env.context.get('default_date_end', False)
+        date_range_type_id = self.env.context.get('default_date_range_type_id',
+                                                  False)
+        location_id = self.env.context.get('default_location_id', False)
+        product_ids = self.env.context.get('product_ids', False)
+        domain = [('type_id', '=', date_range_type_id), '|', '&',
+                  ('date_start', '>=', date_start),
+                  ('date_start', '<=', date_end),
+                  '&',
+                  ('date_end', '>=', date_start),
+                  ('date_end', '<=', date_end)]
+        periods = self.env['date.range'].search(
+            domain)
+        domain = [('type_id', '=', date_range_type_id),
+                  ('date_start', '<=', date_start),
+                  ('date_end', '>=', date_start)]
+        periods |= self.env['date.range'].search(
+            domain)
+        products = self.env['product.product'].browse(product_ids)
+
+        lines = []
+        for product in products:
```

Comment:
so the whole thing is useless if we don't have products, right? Then we should refactor this (and split default values):
```python
    def _default_estimate_ids(self):
        product_ids = self.env.context.get('product_ids', False)
        if not product_ids:
            # Nothing to do
            return None
        date_start = self.env.context.get('default_date_start', False)
        date_end = self.env.context.get('default_date_end', False)
        date_range_type_id = self.env.context.get('default_date_range_type_id',
                                                  False)
        location_id = self.env.context.get('default_location_id', False)
        domain = [('type_id', '=', date_range_type_id), '|', '&',
                  ('date_start', '>=', date_start),
                  ('date_start', '<=', date_end),
                  '&',
                  ('date_end', '>=', date_start),
                  ('date_end', '<=', date_end)]
        periods = self.env['date.range'].search(domain)
        domain = [('type_id', '=', date_range_type_id),
                  ('date_start', '<=', date_start),
                  ('date_end', '>=', date_start)]
        periods |= self.env['date.range'].search(domain)
        products = self.env['product.product'].browse(product_ids)

        lines = []
        for product in products:
            for period in periods:
                estimate = self.env['stock.demand.estimate'].search(
                    [('product_id', '=', product.id),
                     ('date_range_id', '=', period.id),
                     ('location_id', '=', location_id)], limit=1)
                if estimate:
                    uom_id = estimate.product_uom.id
                    uom_qty = estimate.product_uom_qty.id
                    estimate_id = estimate.id
                else:
                    uom_id = product.uom_id.id
                    uom_qty = 0.0
                    estimate_id = None
                lines.append((0, 0, self._get_default_estimate_line(
                    period, product, location_id,
                    uom_id, uom_qty, estimate_id=estimate_id,
                )))
        return lines

    def _get_default_estimate_line(self, period, product,
                                   location_id, uom_id, uom_qty,
                                   estimate_id=None):
        name_y = '{} - {}'.format(product.name, product.uom_id.name)
        if product.default_code:
            name_y += '[{}] {}'.format(product.default_code, name_y)
        values = {
            'value_x': period.name,
            'value_y': name_y,
            'date_range_id': period.id,
            'product_id': product_id,
            'product_uom': uom_id,
            'product_uom_qty': uom_qty
            'location_id': location_id,
            'estimate_id': estimate_id,
        }
        return values
```